File: 11 semester/Games/LR_5/Source/LR_5/Cell.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "Cell.h"

Cell::Cell() {
    State = CreateCellState();
}

bool Cell::IsOpen() const {
    return State->IsOpen;
}

bool Cell::IsBomb() const {
    return dynamic_cast<BombCellState*>(State) != nullptr;
}

void Cell::Attach(IObserver* observer) {
    State->Attach(observer);
}

void Cell::Detach(IObserver* observer) {
    State->Detach(observer);
}

void Cell::Notify() {
    State->Notify();
}

void Cell::Update(IObservable* subject) {
    State->Update(subject);
}

int Cell::Click() {
    return State->Click();
}

bool Cell::RightClick() {
    return State->RightClick();
}

Cell::~Cell() {
    delete State;
}

File: 11 semester/Games/LR_5_unity/Assets/Scripts/Cell.cs
public abstract class Cell : IObservable, IObserver
{
    public CellState CellState { get; set; }

    public Cell()
    {
        CellState = CreateCellState();
    }

    public bool IsOpen
    {
        get
        {
            return CellState.IsOpen;
        }
    }

    public bool IsBomb
    {
        get
        {
            if (CellState is BombCellState)
            {
                return true;
            }
            return false;
        }
    }
    public void Attach(IObserver observer)
    {
        CellState.Attach(observer);
    }

    public void Detach(IObserver observer)
    {
        CellState.Detach(observer);
    }

    public void Notify()
    {
        CellState.Notify();
    }

    public void Update(IObservable subject)
    {
        CellState.Update(subject);
    }
    public int Click()
    {
        return CellState.Click();
    }
    public bool RightClick()
    {
        return CellState.RightClick();
    }

    public abstract CellState CreateCellState();
}